### ISSUE TITLE
helm 3.0 beta2 has same release in one namespace

### ISSUE THREAD

**Issue by wu0407** [helm/helm#6356]
helm install from repo and upgrade from local
```
helm install prometheus-operator  offical-repo/prometheus-operator --namespace=kube-system 
helm upgrade prometheus-operator .  -n kube-system -f values.yaml

helm list -n kube-system                                      
NAME                    NAMESPACE       REVISION        UPDATED                                 STATUS          CHART                    
prometheus-operator     kube-system     16              2019-09-01 12:18:36.589258986 +0800 CST failed          prometheus-operator-6.8.0
prometheus-operator     kube-system     32              2019-09-04 20:57:28.827982696 +0800 CST deployed        prometheus-operator-6.8.0
```

Output of `helm version`:
```
version.BuildInfo{Version:"v3.0.0-beta.2", GitCommit:"26c7338408f8db593f93cd7c963ad56f67f662d4", GitTreeState:"clean", GoVersion:"go1.12.9"}
```

Output of `kubectl version`:
```
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.1", GitCommit:"b7394102d6ef778017f2ca4046abbaa23b88c290", GitTreeState:"clean", BuildDate:"2019-04-08T17:11:31Z", GoVersion:"go1.12.1", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.1", GitCommit:"b7394102d6ef778017f2ca4046abbaa23b88c290", GitTreeState:"clean", BuildDate:"2019-04-08T17:02:58Z", GoVersion:"go1.12.1", Compiler:"gc", Platform:"linux/amd64"}
```

Cloud Provider/Platform (AKS, GKE, Minikube etc.): 
none

**Comment by bacongobbler** [comment]
It looks like `helm list` is just accidentally showing revisions 16 and 32 from the same release ledger. Perhaps the "failed" and "deployed" statuses are significant here.

If you can help determine how to trigger this bug, that would be much appreciated.

**Comment by wu0407** [comment]
helm history
```
REVISION        UPDATED                                 STATUS          CHART                           APP VERSION     DESCRIPTION                                                                                            
1               2019-08-30 22:06:36.936896902 +0800 CST superseded      prometheus-operator-6.7.3       0.31.1          Install complete                                                                                       
2               2019-08-30 22:13:10.385724362 +0800 CST superseded      prometheus-operator-6.7.3       0.31.1          Upgrade complete                                                                                       
3               2019-08-30 23:04:01.538005309 +0800 CST superseded      prometheus-operator-6.7.3       0.31.1          Upgrade complete                                                                                       
4               2019-08-30 23:25:37.447711844 +0800 CST superseded      prometheus-operator-6.7.3       0.31.1          Upgrade complete                                                                                       
5               2019-08-30 23:33:24.533246825 +0800 CST superseded      prometheus-operator-6.7.3       0.31.1          Upgrade complete                                                                                       
6               2019-08-30 23:37:38.931268306 +0800 CST superseded      prometheus-operator-6.7.3       0.31.1          Upgrade complete                                                                                       
7               2019-08-30 23:45:37.158245891 +0800 CST superseded      prometheus-operator-6.7.3       0.31.1          Upgrade complete                                                                                       
8               2019-08-31 00:00:26.17132705 +0800 CST  superseded      prometheus-operator-6.7.3       0.31.1          Upgrade complete                                                                                       
9               2019-08-31 00:15:34.815129264 +0800 CST superseded      prometheus-operator-6.7.3       0.31.1          Upgrade complete                                                                                       
10              2019-08-31 13:43:15.274402082 +0800 CST superseded      prometheus-operator-6.7.3       0.31.1          Upgrade complete                                                                                       
11              2019-08-31 13:47:25.445101129 +0800 CST superseded      prometheus-operator-6.7.4       0.31.1          Upgrade complete                                                                                       
12              2019-09-01 09:55:47.661616337 +0800 CST superseded      prometheus-operator-6.7.4       0.31.1          Upgrade complete                                                                                       
13              2019-09-01 09:57:43.317047258 +0800 CST superseded      prometheus-operator-6.8.0       0.32.0          Upgrade complete                                                                                       
14              2019-09-01 11:43:55.415348925 +0800 CST superseded      prometheus-operator-6.8.0       0.32.0          Upgrade complete                                                                                       
15              2019-09-01 11:51:14.792913904 +0800 CST superseded      prometheus-operator-6.8.0       0.32.0          Upgrade complete                                                                                       
16              2019-09-01 12:18:36.589258986 +0800 CST failed          prometheus-operator-6.8.0       0.32.0          Upgrade "prometheus-operator" failed: no Endpoints with the name "prometheus-operator-kube-proxy" found
17              2019-09-01 12:22:02.66240848 +0800 CST  superseded      prometheus-operator-6.8.0       0.32.0          Upgrade complete                                                                                       
18              2019-09-01 12:27:42.886661544 +0800 CST superseded      prometheus-operator-6.8.0       0.32.0          Upgrade complete                                                                                       
19              2019-09-01 18:24:39.909025984 +0800 CST superseded      prometheus-operator-6.8.0       0.32.0          Upgrade complete                                                                                       
20              2019-09-01 18:43:01.520667625 +0800 CST superseded      prometheus-operator-6.8.0       0.32.0          Upgrade complete                                                                                       
21              2019-09-01 20:03:11.12617428 +0800 CST  superseded      prometheus-operator-6.8.0       0.32.0          Upgrade complete                                                                                       
22              2019-09-02 21:30:11.99995154 +0800 CST  superseded      prometheus-operator-6.8.0       0.32.0          Upgrade complete                                                                                       
23              2019-09-03 23:07:44.19565185 +0800 CST  superseded      prometheus-operator-6.8.0       0.32.0          Upgrade complete                                                                                       
24              2019-09-03 23:28:09.94832683 +0800 CST  superseded      grafana-3.8.7                   6.3.5           Upgrade complete                                                                                       
25              2019-09-03 23:29:23.622010547 +0800 CST superseded      prometheus-operator-6.8.0       0.32.0          Upgrade complete                                                                                       
26              2019-09-03 23:33:03.021662324 +0800 CST superseded      prometheus-operator-6.8.0       0.32.0          Upgrade complete                                                                                       
27              2019-09-04 19:49:42.041930887 +0800 CST superseded      prometheus-operator-6.8.0       0.32.0          Upgrade complete                                                                                       
28              2019-09-04 19:52:14.927278375 +0800 CST superseded      prometheus-operator-6.8.0       0.32.0          Upgrade complete                                                                                       
29              2019-09-04 20:49:31.667397909 +0800 CST superseded      prometheus-operator-6.8.0       0.32.0          Upgrade complete                                                                                       
30              2019-09-04 20:53:33.018812982 +0800 CST superseded      prometheus-operator-6.8.0       0.32.0          Upgrade complete                                                                                       
31              2019-09-04 20:56:42.500947009 +0800 CST superseded      prometheus-operator-6.8.0       0.32.0          Upgrade complete                                                                                       
32              2019-09-04 20:57:28.827982696 +0800 CST superseded      prometheus-operator-6.8.0       0.32.0          Upgrade complete                                                                                       
33              2019-09-05 13:44:30.836485601 +0800 CST superseded      prometheus-operator-6.8.0       0.32.0          Upgrade complete                                                                                       
34              2019-09-05 15:06:47.714682612 +0800 CST superseded      prometheus-operator-6.8.0       0.32.0          Upgrade complete                                                                                       
35              2019-09-05 19:00:44.537315956 +0800 CST superseded      prometheus-operator-6.8.0       0.32.0          Upgrade complete                                                                                       
36              2019-09-05 19:06:10.919309686 +0800 CST superseded      prometheus-operator-6.8.0       0.32.0          Upgrade complete                                                                                       
37              2019-09-05 22:40:40.817256468 +0800 CST superseded      prometheus-operator-6.8.0       0.32.0          Upgrade complete                                                                                       
38              2019-09-05 22:48:43.035534547 +0800 CST superseded      prometheus-operator-6.8.0       0.32.0          Upgrade complete                                                                                       
39              2019-09-05 23:09:06.058808444 +0800 CST superseded      prometheus-operator-6.8.0       0.32.0          Upgrade complete                                                                                       
40              2019-09-06 00:11:21.961810874 +0800 CST superseded      prometheus-operator-6.8.0       0.32.0          Upgrade complete                                                                                       
41              2019-09-06 00:27:13.440355637 +0800 CST superseded      prometheus-operator-6.8.0       0.32.0          Upgrade complete                                                                                       
42              2019-09-06 09:25:46.866824653 +0800 CST superseded      prometheus-operator-6.8.0       0.32.0          Upgrade complete                                                                                       
43              2019-09-06 09:34:43.231632172 +0800 CST superseded      prometheus-operator-6.8.0       0.32.0          Upgrade complete                                                                                       
44              2019-09-06 10:46:34.992708367 +0800 CST superseded      prometheus-operator-6.8.0       0.32.0          Upgrade complete                                                                                       
45              2019-09-06 11:19:54.967575516 +0800 CST superseded      prometheus-operator-6.8.0       0.32.0          Upgrade complete                                                                                       
46              2019-09-06 11:44:30.443543231 +0800 CST superseded      prometheus-operator-6.8.0       0.32.0          Upgrade complete                                                                                       
47              2019-09-06 14:01:53.56933603 +0800 CST  deployed        prometheus-operator-6.8.0       0.32.0          Upgrade complete     
```

**Comment by karuppiah7890** [comment]
Able to reproduce the issue. Install this chart

[issue-6356-chart.zip](https://github.com/helm/helm/files/3586307/issue-6356-chart.zip)

Upgrade with this chart
[issue-6356-chart-new.zip](https://github.com/helm/helm/files/3586306/issue-6356-chart-new.zip)

Upgrade will fail due to immutable field error. And running helm list shows this

```
$ helm ls
NAME    NAMESPACE       REVISION        UPDATED                                 STATUS          CHART
dummy   default         1               2019-09-07 11:50:04.078966 +0530 IST    deployed        dummy-chart-0.1.0
dummy   default         2               2019-09-07 11:52:18.867486 +0530 IST    failed          dummy-chart-0.1.0
```

Doing another upgrade with same `issue-6356-chart-new` leads to another error and helm list shows this

```
$ helm ls
NAME    NAMESPACE       REVISION        UPDATED                                 STATUS          CHART
dummy   default         1               2019-09-07 11:50:04.078966 +0530 IST    deployed        dummy-chart-0.1.0
dummy   default         2               2019-09-07 11:52:18.867486 +0530 IST    failed          dummy-chart-0.1.0
dummy   default         3               2019-09-07 11:55:29.931651 +0530 IST    failed          dummy-chart-0.1.0
```

basically all failed releases are coming. Finally upgrading with `issue-6356-chart` passes and helm list shows this

```
$ helm ls
NAME    NAMESPACE       REVISION        UPDATED                                 STATUS          CHART
dummy   default         2               2019-09-07 11:52:18.867486 +0530 IST    failed          dummy-chart-0.1.0
dummy   default         3               2019-09-07 11:55:29.931651 +0530 IST    failed          dummy-chart-0.1.0
dummy   default         4               2019-09-07 11:57:52.858755 +0530 IST    deployed        dummy-chart-0.1.0
```

Checking it out. Will raise a PR if I find the issue 😄

**Comment by karuppiah7890** [comment]
This is what is currently happening - we take all the releases information and do the filtering based on state, for which the default is `deployed` or `failed`. The thing is, when we get releases information, we get all the revisions of all the releases and then do the filtering, so, given a release name, multiple revisions will show up. And this happens for all the release storage drivers - Secrets, ConfigMaps, In-Memory, where helm could possibly store releases.

I was thinking we could put a fix in the storage driver code, where when we say list and filter based on a passed function. Like for secrets, inside this function

https://github.com/helm/helm/blob/895e9192d440e144ead0796fd8c1db81b47ef1c6/pkg/storage/driver/secrets.go#L80

Before doing filtering, we can find the latest revision of each release and get rid of the old revisions. In helm v2, I think there's some similar code to do this, over here

https://github.com/helm/helm/blob/8fcc438b67de564a3dd9255ec9c1634229b764c0/cmd/helm/list.go#L197

One thing to note is, we need to implement the removal of old revisions in the `List` method of all the storage drivers. This is something to keep in mind, and we may also need to put a comment on `List` method saying that it lists all latest revisions of releases, and does not include old ones.

Another solution that I have is - to finish off filtering based on status first, then filter out the old revisions, which doesn't exactly make sense as there will be lots of old revisions for each release, so there would be unnecessary filtering computation being done for it and then later they will get filtered out. Doesn't sound performant though. I'm not sure if there are any advantages that I'm overlooking here though, like library API etc, but performance is still important and this doesn't look performant

I'm willing to raise a PR for the fix based on the solution we choose. 

@bacongobbler What are your thoughts on this? 😄 Any other solutions that you have in mind for the fix?

**Comment by bacongobbler** [comment]
That sounds about right - it's likely something in the current filters used by `helm list` that's causing failed releases to show up.

**Comment by karuppiah7890** [comment]
Cool, I'll write some tests for this and fix it 😄  the issue also shows up differently in `helm ls -a` command, even when there are no failed releases.

**Comment by karuppiah7890** [comment]
Choosing this solution for the PR : 
> we could put a fix in the storage driver code, where when we say list and filter based on a passed function.
> Before doing filtering, we can find the latest revision of each release and get rid of the old revisions. 
> One thing to note is, we need to implement the removal of old revisions in the List method of all the storage drivers.

**Comment by karuppiah7890** [comment]
I have raised a PR for this! @bacongobbler

**Comment by bacongobbler** [comment]
fixed in #6479
